llvm/include/llvm/IR/DataLayout.h

```diff
@@ -164,6 +165,8 @@ class DataLayout {
   /// well-defined bitwise representation.
   SmallVector<unsigned, 8> NonIntegralAddressSpaces;
 
+  DenseMap<uint64_t, uint64_t> AddrSpaceToNonZeroValueMap;
```

Comment:
updated in the latest.

Reply:
I'm not aware of any fat pointers where the *address* part is >64 bits.  I agree with this change though, because RISC-V does have a 128-bit address variant and some folks are working on Linux support for it.

In CHERI, we made the decision that an all-zeroes bit pattern would be the canonical null pointer because, although C requires only that an integer *constant* expression (or one that is converted to a pointer) is null, C programmers *really* don't pay attention to the 'constant' bit of that.  Being able to `memset` a struct and have it full of zero integers and null pointers is important.  Assuming that calloc'd memory is full of null pointers is important.  And, broadly, the following is *super* confusing:

```c
const int null = 0;
int not_null = 0;
// This passes
assert(null == not_null);
// This fails
assert((void*)null == (void*)not_null);
```

Oh, and whether the second one fails may depend on optimisation levels.  

I still remain of the opinion that trying to make the null value anything other than zero is a terrible idea and the correct fix for AMDGPU is to make INTTOPTR and PTRTOINT subtract and add 1, respectively, so converying a 0 integer to a pointer yields a pointer with a -1 value.

Reply:
Unless we can start folding out expressions based on inttoptr, changing the lowering of inttoptr/ptrtoint only causes more problems. We do not have a correctness representational problem today (modulo bugs).

We have an optimization issue, because we end up with cases like #58617 where obviously false pointer comparisons can't be deleted. The IR needs to be aware of the non-0 behavior, rather than just ignoring it for globals/functions/nonnull decorated pointers